File: Modelo Columnar/c# columnar/consultas.cs
using Cassandra;
using Cassandra.Data.Linq;
using CassandraCSharpExample;
using System;
using System.Linq;


namespace columnas
{
    class consultas
    {
        static void Main(string[] args)
        {
            // Configura la conexión a tu clúster de Cassandra
            var cluster = Cluster.Builder()
                .AddContactPoint("127.0.0.1")
                .Build();


            var session = cluster.Connect("dbcolumnas");

            /*
             //¿Cuántos tipos de población hay y cuales son?

            var consulta = "SELECT tipopoblacion FROM Programa"; // Ajusta la consulta según tu modelo

            var resultados = session.Execute(consulta);

            // Crear un conjunto para almacenar tipos de población únicos
            var tiposDePoblacionUnicos = new HashSet<string>();

            foreach (var fila in resultados)
            {
                // Accede al valor de "tipopoblacion"
                var tipoPoblacion = fila.GetValue<string>("tipopoblacion");

                // Agrega el tipo de población al conjunto si es diferente de nulo
                if (!string.IsNullOrEmpty(tipoPoblacion))
                {
                    tiposDePoblacionUnicos.Add(tipoPoblacion);
                }
            }

            // Imprime la cantidad de tipos de población y la lista de tipos
            Console.WriteLine($"Cantidad de tipos de población: {tiposDePoblacionUnicos.Count}");
            Console.WriteLine("Tipos de población:");

            foreach (var tipo in tiposDePoblacionUnicos)
            {
                Console.WriteLine(tipo);
            }
            */


            /*
              //  ¿Cuál es el rango más alto de beneficio consolidado asignado?

              var consulta = "SELECT rangobeneficioconsolidadoasignado FROM Programa";

              var resultados = session.Execute(consulta);

              // Almacena los rangos en una lista
              var rangos = new List<string>();

              foreach (var fila in resultados)
              {
                  var rango = fila.GetValue<string>("rangobeneficioconsolidadoasignado");
                  rangos.Add(rango);
              }

              // Encuentra el rango más alto en la lista
              var rangoMasAlto = rangos.Count > 0 ? rangos.Max() : "No hay datos";

              Console.WriteLine($"Rango más alto de beneficio consolidado asignado: {rangoMasAlto}");

              */

            /*
            // programas con el limite mas alto de beneficiario cuando el tipo de asignacion es MONETARIO


            var consulta = "SELECT id, tipoasignacionbeneficio, cantidaddebeneficiarios FROM Programa WHERE tipoasignacionbeneficio = 'MONETARIO' ALLOW FILTERING";

            // Ejecutar la consulta en Cassandra
            var resultados = session.Execute(consulta);

           // Crear una lista para almacenar programas como tuplas (ID, Tipo de Asignación, Cantidad de Beneficiarios)
           var programas = new List<Tuple<Guid, string, int>>();

           // Iterar a través de los resultados de la consulta
           foreach (var fila in resultados)
           {
              // Obtener el ID, Tipo de Asignación y Cantidad de Beneficiarios de cada fila
             var id = fila.GetValue<Guid>("id");
             var tipoAsignacion = fila.GetValue<string>("tipoasignacionbeneficio");
             var cantidad = fila.GetValue<int>("cantidaddebeneficiarios");

             // Agregar los datos como una tupla a la lista de programas
             programas.Add(new Tuple<Guid, string, int>(id, tipoAsignacion, cantidad));
           }

             // Ordenar la lista de programas por la cantidad de beneficiarios en orden descendente
             programas.Sort((a, b) => b.Item3.CompareTo(a.Item3));

              // Encontrar el límite más alto de beneficiarios
              var limiteMasAlto = programas.Any() ? programas[0].Item3 : 0;

              // Mostrar el límite más alto de beneficiarios en la consola
              Console.WriteLine($"El límite más alto de beneficiarios es: {limiteMasAlto}");

              // Verificar si hay programas con el límite más alto
             if (programas.Any())
             {
               // Mostrar programas con el límite más alto de beneficiarios y sus detalles
               Console.WriteLine("Programas con el límite más alto de beneficiarios:");

                foreach (var programa in programas)
                 {
                    Console.WriteLine($"ID: {programa.Item1}, Tipo de Asignación: {programa.Item2}, Cantidad de Beneficiarios: {programa.Item3}");
                 }
             }
             else
             {
                // Mostrar un mensaje si no se encontraron programas con asignación 'MONETARIO'
                Console.WriteLine("No se encontraron programas con el tipo de asignación 'MONETARIO'.");
             }

             */

            /*
            // columnas de la tabla Programa
            var consulta = "SELECT * FROM Programa LIMIT 1";
            var resultados = session.Execute(consulta);

            var metadata = resultados.Columns;

            foreach (var columna in metadata)
            {
                Console.WriteLine(columna.Name);
            }

            */




            /*
            //¿Cuántos beneficiarios se encuentran bancarizados?

            var consulta = "SELECT COUNT(*) AS cantidad_bancarizados FROM Beneficiario WHERE bancarizado = 'SI' ALLOW FILTERING";

            var resultados = session.Execute(consulta);

            foreach (var fila in resultados)
            {
                var cantidadBancarizados = fila.GetValue<long>("cantidad_bancarizados");

                Console.WriteLine($"Cantidad de beneficiarios bancarizados: {cantidadBancarizados}");
            }

            */

            
            // ¿Cuánto beneficiarios hay por cada género?
     
            var consulta = "SELECT genero FROM Beneficiario";

            var resultados = session.Execute(consulta);

            // Almacena los géneros en una lista
            var generos = new List<string>();

            foreach (var fila in resultados)
            {
                var genero = fila.GetValue<string>("genero");
                generos.Add(genero);
            }

            // Realiza la agregación de la cantidad de beneficiarios por género
            var conteoPorGenero = new Dictionary<string, int>();

            foreach (var genero in generos)
            {
                if (conteoPorGenero.ContainsKey(genero))
                {
                    conteoPorGenero[genero]++;
                }
                else
                {
                    conteoPorGenero[genero] = 1;
                }
            }

            // Muestra la cantidad de beneficiarios por género
            Console.WriteLine("Cantidad de beneficiarios por género:");

            foreach (var kvp in conteoPorGenero)
            {
                Console.WriteLine($"Género: {kvp.Key}, Cantidad de beneficiarios: {kvp.Value}");
            }
            


            /*
            // nombre de los departamentos?

                         // Define la tabla de departamentos
                         var departamentoTable = new Table<Departamento>(session);

                         // Realiza la consulta para obtener los nombres de los departamentos
                         var query = departamentoTable.Select(d => new { d.nombredepartamentoatencion });

                         // Ejecuta la consulta y obtén los resultados
                         var results = query.Execute();

                         // Imprime los nombres de los departamentos
                         foreach (var result in results)
                         {
                             Console.WriteLine(result.nombredepartamentoatencion);
                         }

              */


            /* //tabla departamento
                // Define la tabla de departamentos
                var departamentoTable = new Table<Departamento>(session);

                // Realiza la consulta para obtener todos los atributos de los departamentos
                var query = departamentoTable.Select(d => d);

                // Ejecuta la consulta y obtén los resultados
                var results = query.Execute();

                        // Imprime todos los atributos de los departamentos
                        foreach (var result in results)
                        {
                            Console.WriteLine($"ID: {result.id}, Código Departamento: {result.codigodepartamentoatencion}, Nombre Departamento: {result.nombredepartamentoatencion}");
                        }

                        */
            // Cerrar la conexión
            cluster.Dispose();

            Console.WriteLine("Consulta finalizada.");
            Console.ReadKey();
        }
    }
}

File: Modelo Columnar/c# columnar/consultas.py
from cassandra.cluster import Cluster

# Configura la conexión a tu clúster de Cassandra
cluster = Cluster(['127.0.0.1']) 
session = cluster.connect('dbcolumnar')  


# ¿Cuántos tipos de población hay y cuales son?

consulta = "SELECT tipopoblacion FROM Programa"

resultados = session.execute(consulta)

# Crear un conjunto para almacenar tipos de población únicos
tiposDePoblacionUnicos = set()

for fila in resultados:
    # Accede al valor de "tipopoblacion"
    tipoPoblacion = fila.tipopoblacion

    # Agrega el tipo de población al conjunto si es diferente de nulo
    if tipoPoblacion is not None:
        tiposDePoblacionUnicos.add(tipoPoblacion)

# Imprime la cantidad de tipos de población y la lista de tipos
print(f"Cantidad de tipos de población: {len(tiposDePoblacionUnicos)}")
print("Tipos de población:")
for tipo in tiposDePoblacionUnicos:
    print(tipo)




"""
# ¿Cuál es el rango más alto de beneficio consolidado asignado?

consulta = "SELECT rangobeneficioconsolidadoasignado FROM Programa"

resultados = session.execute(consulta)

# Almacena los rangos en una lista
rangos = []

for fila in resultados:
    rango = fila.rangobeneficioconsolidadoasignado
    rangos.append(rango)

# Encuentra el rango más alto en la lista
rangoMasAlto = max(rangos, default="No hay datos")

print(f"Rango más alto de beneficio consolidado asignado: {rangoMasAlto}")
"""



"""
# programa con el limite mas alto de beneficiario cuando el tipo de asignacion es ND

consulta = "SELECT id, tipoasignacionbeneficio, cantidaddebeneficiarios FROM Programa WHERE tipoasignacionbeneficio = 'ND' ALLOW FILTERING"

# Ejecutar la consulta en Cassandra
resultados = session.execute(consulta)

# Crear una lista para almacenar programas como tuplas (ID, Tipo de Asignación, Cantidad de Beneficiarios)
programas = []

# Iterar a través de los resultados de la consulta
for fila in resultados:
    # Obtener el ID, Tipo de Asignación y Cantidad de Beneficiarios de cada fila
    id_programa = fila.id
    tipo_asignacion = fila.tipoasignacionbeneficio
    cantidad_beneficiarios = fila.cantidaddebeneficiarios

    # Agregar los datos como una tupla a la lista de programas
    programas.append((id_programa, tipo_asignacion, cantidad_beneficiarios))

# Ordenar la lista de programas por la cantidad de beneficiarios en orden descendente
programas.sort(key=lambda x: x[2], reverse=True)

# Encontrar el límite más alto de beneficiarios
limiteMasAlto = programas[0][2] if programas else 0

print(f"El límite más alto de beneficiarios es: {limiteMasAlto}")
 
 """



'''
# ¿Cuántos beneficiarios se encuentran bancarizados?
consulta = "SELECT COUNT(*) AS cantidad_bancarizados FROM Beneficiario WHERE bancarizado = 'SI' ALLOW FILTERING"

resultados = session.execute(consulta)

for fila in resultados:
    cantidadBancarizados = fila.cantidad_bancarizados

print(f"Cantidad de beneficiarios bancarizados: {cantidadBancarizados}")

'''


'''
# ¿Cuántos beneficiarios hay por cada género?
consulta = "SELECT genero FROM Beneficiario"

resultados = session.execute(consulta)

# Almacena los géneros en una lista
generos = []

for fila in resultados:
    genero = fila.genero
    generos.append(genero)

# Realiza la agregación de la cantidad de beneficiarios por género
conteoPorGenero = {}

for genero in generos:
    if genero in conteoPorGenero:
        conteoPorGenero[genero] += 1
    else:
        conteoPorGenero[genero] = 1

# Muestra la cantidad de beneficiarios por género
print("Cantidad de beneficiarios por género:")
for genero, cantidad in conteoPorGenero.items():
    print(f"Género: {genero}, Cantidad de beneficiarios: {cantidad}")

'''


'''

# nombre de los primeros 5 departamentos?

# Define la tabla de departamentos
departamentoTable = session.execute("SELECT * FROM Departamento LIMIT 5")

# Imprime los nombres de los departamentos
for row in departamentoTable:
    print(row.nombredepartamentoatencion)
'''


'''
# tabla departamento

# Define la tabla de departamentos
departamentoTable = session.execute("SELECT * FROM Departamento")

# Imprime todos los atributos de los departamentos
for row in departamentoTable:
    print(f"ID: {row.id}, Código Departamento: {row.codigodepartamentoatencion}, Nombre Departamento: {row.nombredepartamentoatencion}")
'''

"""
# columnas de la tabla Programa
consulta = "SELECT * FROM Programa "
resultados = session.execute(consulta)

data = resultados.column_names

for columna in data:
    print(columna)
"""


"""
# Consulta Municipios que pertenecen al departamentod del cesar

# Consulta para obtener los municipios del departamento del Cesar
consulta = "SELECT  nombremunicipioatencion FROM Municipio WHERE codigodepartamentoatencion = '20' ALLOW FILTERING"

# Ejecuta la consulta
result = session.execute(consulta)

# Imprime los nombres de los municipios
print("Municipios del departamento del Cesar:")
for row in result:
    print(row.nombremunicipioatencion)


"""



# Cerrar la conexión
cluster.shutdown()
print("Consulta finalizada.")
